pkg/controller/priorityqueue/priorityqueue.go

```diff
@@ -195,35 +195,41 @@ func (w *priorityqueue[T]) spin() {
 			w.lockedLock.Lock()
 			defer w.lockedLock.Unlock()
 
-			w.queue.Ascend(func(item *item[T]) bool {
-				if w.waiters.Load() == 0 { // no waiters, return as we can not hand anything out anyways
-					return false
-				}
+			for continueLoop := true; continueLoop; {
+				continueLoop = false
+				w.queue.Ascend(func(item *item[T]) bool {
+					if w.waiters.Load() == 0 { // no waiters, return as we can not hand anything out anyways
+						return false
+					}
+
+					// No next element we can process
+					if item.readyAt != nil && item.readyAt.After(w.now()) {
+						readyAt := item.readyAt.Sub(w.now())
+						if readyAt <= 0 { // Toctou race with the above check
+							readyAt = 1
+						}
+						nextReady = w.tick(readyAt)
+						return false
+					}
 
-				// No next element we can process
-				if item.readyAt != nil && item.readyAt.After(w.now()) {
-					readyAt := item.readyAt.Sub(w.now())
-					if readyAt <= 0 { // Toctou race with the above check
-						readyAt = 1
+					// Item is locked, we can not hand it out
+					if w.locked.Has(item.key) {
+						return true
 					}
-					nextReady = w.tick(readyAt)
+
+					w.metrics.get(item.key)
+					w.locked.Insert(item.key)
+					w.waiters.Add(-1)
+					delete(w.items, item.key)
+					w.queue.Delete(item)
+					w.get <- *item
+
+					// Return false because continuing with Ascend after deleting an item
+					// can lead to panics within Ascend.
+					continueLoop = true
 					return false
```

Comment:
Not a super strong opinion but how do you feel about instead appending to a `toDelete` slice in `Ascend` and then calling `Delete` after being  done with `Ascend`? It should be safe because we are holding the lock so a concurrent routine seeing the item when it is supposed to be deleted shouldn't be possible.

The reason is that even if it works this way now, I don't think manipluting the tree while iterating is an expected usage, even if it seems to work now there could be more bugs or it could stop working in a future version of the lib.

Reply:
I ended up using that approach + your testcase in https://github.com/kubernetes-sigs/controller-runtime/pull/3060 as this issue made CI fail there, hope that is okay

Reply:
Yup. You're right, seems safer to just not delete within Ascend. I'll take a look at your PR. Thx for taking this over